### PR TITLE
[ENG-1552] Make departments work

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

Create an endpoint where the institutional admin can get sorted and filter department info.

## Changes

- add InstitutionDepartmentList view and serializer class
- add tests with ES setup for travis
- unskip ES test that was skipped
- add Mock Queryset so raw ES results can be filtered

## QA Notes

No Migration for this, but QA and devs must coordinate how data is populated for this. filtering. permissions and other API behavior should be tested.

Sample response:
`http://localhost:8000/v2/institutions/cos/departments/`
```json
{
    "data": [
        {
            "id": "cos",
            "type": "institution-departments",
            "attributes": {
                "name": "Section 9",
                "number_of_users": 1
            },
            "links": {
                "self": "http://localhost:8000/v2/institutions/cos/departments/"
            }
        }
    ],
    "meta": {
        "total": 1,
        "per_page": 10,
        "version": "2.20"
    },
    "links": {
        "self": "http://localhost:8000/v2/institutions/cos/departments/",
        "first": null,
        "last": null,
        "prev": null,
        "next": null
    }
}
```

## Documentation

No additional docs, because the endpoint is restricted to institutional admins.

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1552

### CODE REVIEW

File: api_tests/metrics/test_preprint_metrics.py [hunk 1]
```diff
@@ -116,17 +116,9 @@ def test_custom_metric_misformed_query(self, mock_execute, app, user, base_url):
         assert res.status_code == 400
         assert res.json['errors'][0]['detail'] == 'Misformed elasticsearch query.'
 
-    @pytest.mark.skip('Return results will be entirely mocked so does not make a lot of sense to run on travis.')
+    @pytest.mark.es
```
Comment (Johnetordoff) [inline api_tests/metrics/test_preprint_metrics.py]: Drive-by change, because ES6 testing has been figured out now.

File: pytest.ini [hunk 2]
```diff
@@ -1,5 +1,5 @@
 [pytest]
-addopts = --ds=osf_tests.settings --tb=short --reuse-db --allow-hosts=127.0.0.1
+addopts = --ds=osf_tests.settings --tb=short --reuse-db --allow-hosts=127.0.0.1,192.168.168.167
```
Comment (Johnetordoff) [inline pytest.ini]: lets us access ES container in pytest.

File: osf/permissions.py [hunk 3]
```diff
@@ -0,0 +1 @@
+INSTITUTION_ADMIN = 'institution_admin'
```
Comment (UdayVarkhedkar) [inline osf/permissions.py]: Why is this defined in a new file as opposed to osf/utils/permissions.py?

Comment (Johnetordoff) [reply osf/permissions.py]: good point I just missed it.

Comment (corbinSanders) [reply osf/permissions.py]: Drive by: @Johnetordoff, here is the PR with the permissions for institutional users. https://github.com/CenterForOpenScience/osf.io/pull/9297/files

Comment (Johnetordoff) [reply osf/permissions.py]: Ah Ok everyone on the same page, pull and work off this #9297 for permissions.

File: api/institutions/views.py [hunk 4]
```diff
@@ -366,3 +377,64 @@ def create(self, *args, **kwargs):
...
+    required_read_scopes = [CoreScopes.INSTITUTION_READ_METRICS]
+    required_write_scopes = [CoreScopes.NULL]
+
+    view_category = 'institutions'
+    view_name = 'institution-department-metrics'
+
+    ordering = ('-number_of_users', 'name',)
+
+    @classmethod
+    def _get_departments_counts(cls, current_user_data: list) -> dict:
```
Comment (UdayVarkhedkar) [inline api/institutions/views.py]: I don't believe summing the project counts of users in a given department will work. There may be multiple users in the same department working on the same project and I believe that should count as 1 project for the department.

Comment (UdayVarkhedkar) [reply api/institutions/views.py]: er- actually, perhaps not. Depends on if the count for a department is the number of distinct projects associated with a given department (through users) or the total number of projects for users associated with a given department.

Comment (Johnetordoff) [reply api/institutions/views.py]: Unless I got the recs wrong, this just counts the number of users in each department, no project counts involved. Return json should look like [this](https://github.com/CenterForOpenScience/osf.io/pull/9318/files#diff-11dee4e28adbf356065341776ce922b7R110-R133)

Comment (UdayVarkhedkar) [reply api/institutions/views.py]: Ah looks like you're right, sorry about that.

File: osf/metrics.py [hunk 5]
```diff
@@ -180,3 +181,64 @@ class PreprintView(BasePreprintMetric):
...
+                                            'order': 'desc'
+                                        }
+                                    }
+                                ]
+                            }
+                        }
+                    }
+                }
+            }
+        })
```
Comment (brianjgeiger) [inline osf/metrics.py]: You should be able to ask ES directly for a list of departments for an institution aggregating the number of people in that department. There should be no need to loop through a list of folks or anything like that. Let ES work for you.

Comment (Johnetordoff) [reply osf/metrics.py]: The problem with this is we don't want to sum every department down the time series, we just the most recent sums for that day. I'd argue this should not use ES at all, as we have all the current department data in postgres as department is a model attribute.

Comment (Johnetordoff) [reply osf/metrics.py]: @brianjgeiger Courtney gave me a great query, so finishing this up with ES only should be no problem, I'll do that, but I maintain the postgress would also work as users have no way of checking department membership at a specific date.

File: osf/metrics.py [hunk 6]
```diff
@@ -196,6 +197,60 @@ class Index:
...
+                        'departments': {
+                            'terms': {
+                                'field': 'department',
+                                'missing': 'N/A',
+                                'size': 250
+                            },
+                            'aggs': {
+                                'users': {
+                                    'terms': {
+                                        'field': 'user_id'
```
Comment (aaxelb) [inline osf/metrics.py]: i get a 500 error for this query locally:
> elasticsearch.exceptions.RequestError: RequestError(400, 'search_phase_execution_exception', 'Fielddata is disabled on text fields by default. Set fielddata=true on [user_id] in order to load fielddata in memory by uninverting the inverted index. Note that this can however use significant memory. Alternatively use a keyword field instead.')

checked out the mapping for the `user_id` field:
```
          "user_id" : {
            "type" : "text",
            "fields" : {
              "keyword" : {
                "type" : "keyword",
                "ignore_above" : 256
              }
            }
          }
```
and looks like, while `user_id` is a text field, there's also a `user_id.keyword` field that we could use instead.

this change fixed the error for me
```suggestion
                                        'field': 'user_id.keyword'
```
but i'm also curious why this happened for me but not you -- what's your index mapping like? (see http://localhost:9201/osf_userinstitutionprojectcounts_2020/?pretty)

File: osf/metrics.py [hunk 7]
```diff
@@ -196,6 +197,60 @@ class Index:
             'refresh_interval': '1s',
         }
 
+    class Meta:
+        source = metrics.MetaField(enabled=True)
+
+    @classmethod
+    def filter_institution(cls, institution):
+        return cls.search().filter('match', institution_id=institution._id)
```
Comment (aaxelb) [inline osf/metrics.py]: given the text/keyword thing below, it might be faster/better-practice to filter on the indexed-as-keyword `institution_id.keyword` field, instead of the indexed-as-full-text `institution_id` field

File: osf/metrics.py (hunk 6 above)
Comment (aaxelb) [reply osf/metrics.py]: disregard, this was my local mappings being out-of-date -- in the current mappings `user_id` is of type `keyword`

File: osf/metrics.py (hunk 7 above)
Comment (aaxelb) [reply osf/metrics.py]: disregard, this was my local mappings being out-of-date -- in the current mappings `user_id` is of type `keyword`
